blog/2024-10-25-object-notifications/index.mdx

```diff
@@ -0,0 +1,132 @@
+
+<center>
+  <small>
+    <em>
+      Autumn trees on a dusty road in Magoebaskloof, South Africa. Photo by
+      Garren Smith, iPhone 13 Pro.
+    </em>
+  </small>
+</center>
+
+# How We Built Object Notifications in Tigris
+
+Tigris now supports object notifications! Object notifications are how you
+receive events every time something changes in a bucket. Think of it as your
+bucket's way of saying "Hey, something happened! Come check it out!",
+much like the [inotify](https://en.wikipedia.org/wiki/Inotify) subsystem in Linux.
+These notifications can be helpful for keeping track of what's going on in your
+application.
+
+## Use Case: Automatic Image Processing
+
+A simple use case for object notifications is to automatically process images.
+Imagine you're building a photo-sharing app. Every time a user uploads a new
+picture, you want to automatically generate a thumbnail and maybe even run it
+through an AI to detect any inappropriate content. With object notifications,
+this becomes a breeze!
+
+1. User uploads an image to your Tigris bucket.
+2. Tigris sends a notification to your webhook.
+3. Your server receives the notification and springs into action.
+4. It downloads the new image, creates a thumbnail, and runs it through an AI
+   check.
+5. The processed image and its metadata are saved back to Tigris.
+
+All of this happens automatically, triggered by that initial upload.
+
+## Behind the Scenes: Building Object Notifications
+
+Now, let's pull back the curtain and see how we built this feature and a few
+tricky situations we had to handle. Grab your hard hat, because we're going on a
+little tour of Tigris's inner workings!
+
+
+Tigris isn't just any object store – it's a global object store. This means that
+objects can be changed in multiple regions around the world. This makes them
+available in multiple regions, always ready when you need them. But means we
+need a way of keeping track of all the changes for the same object. This is
+where replication comes in.
+
+### Replication: Keeping Everyone in the Loop
+
+To make sure everything stays in sync, we replicate metadata changes to multiple
+regions. This ensures high availability and improved redundancy of our objects.
+
+The caveat to this is that Replication is a background task, and the speed at
+which an object is replicated from one region to another can be affect by many
+external factors.
+
+To solve this, when a change is received at a region it looks at the
+`Last Modified` timestamp of the metadata to determine if the change is new and
+needs to be applied or if the region has already seen a newer change. It will
+discard the change if it is old.
+
+### The Object Notification Hub
+
+When object notifications are enabled for a bucket, we assign one region to be
+the object notification hub. This region gets the important job of keeping track
+of all the changes. We create a a special index in that regions FoundationDB for
```

Comment:
`We create a a special index in that regions FoundationDB for
this. It is very similar to a secondary index.` --> `We create a special index which is very similar to a secondary index.`